master/pyext/src/validation/__init__.py

```diff
@@ -12,527 +12,682 @@
-            empty_chain_dict[el._id]=[]
+    def get_empty_chain_dict(self) -> dict:
+        empty_chain_dict = defaultdict()
+        for ind, el in enumerate(self.system.asym_units):
+            empty_chain_dict[el._id] = []
         return empty_chain_dict
 
-    def get_chain_subunit_dict(self)->dict:
+    def get_chain_subunit_dict(self) -> dict:
         """ Get chains of subunits"""
-        chain_subunit_dict=defaultdict()
-        for ind,el in enumerate(self.system.asym_units):
-            chain_subunit_dict[el._id]=el.details.split('.')[0]
+        chain_subunit_dict = defaultdict()
+        for ind, el in enumerate(self.system.asym_units):
+            chain_subunit_dict[el._id] = el.details.split('.')[0]
         return chain_subunit_dict
 
-    def get_residues_subunit_dict(self)->dict:
+    def get_residues_subunit_dict(self) -> dict:
         """Get residues of chains in subunits"""
-        residues_subunit_dict=defaultdict()
-        for _,el in enumerate(self.system.asym_units):
-            print (self.get_residues(el))
-            residues_subunit_dict[el._id]=self.get_residues(el)
+        residues_subunit_dict = defaultdict()
+        for el in self.system.asym_units:
+            residues_subunit_dict[el._id] = self.get_residues(el)
         return residues_subunit_dict
 
-    def get_software_length(self)->int:
-        lists=self.system.software
+    def get_software_length(self) -> int:
+        lists = self.system.software
         if lists is None:
             return 0
         else:
             return len(lists)
 
-    def get_software_comp (self)->dict:
+    def get_software_comp(self) -> dict:
         """get software composition to write out as a table"""
-        software_comp={'ID':[],'Software name':[],'Software version':[],'Software classification':[],'Software location':[]}
-        lists=self.system.software
-        if len(lists)>0:
-            for _ in lists:
-                software_comp['ID'].append(_._id)
-                software_comp['Software name'].append(_.name)
-                software_comp['Software location'].append(_.location)
-                if str(_.version) == '?':
-                    vers='None'
+        software_comp = {'ID': [], 'Software name': [], 'Software version': [
+        ], 'Software classification': [], 'Software location': []}
+        lists = self.system.software
+        self.read_all_references()
+        if len(lists) > 0:
+            for software in lists:
+                software_comp['ID'].append(software._id)
+                ref_name = software.name.lower()
+                ref_tot = '<a href="' + \
+                    self.ref_link[ref_name]+'">'+software.name+"</a>"
+                ref_loc = '<a href="'+software.location+'">'+software.location+"</a>"
+                software_comp['Software name'].append(ref_tot)
+                software_comp['Software location'].append(ref_loc)
+                if str(software.version) == ihm.unknown:
```

Comment:
You probably just want `if software.version in (None, ihm.unknown):` here. `ihm.unknown` is not a string in modern python-ihm (even though it looks like one) so won't compare equal to "?". `None` covers the case where the user put in `.` for the version.